sklearn/tree/_splitter.pyx

```diff
@@ -373,6 +393,10 @@ cdef class BestSplitter(BaseDenseSplitter):
 
                 self.criterion.update(current.pos)
 
+                # Reject if monotonicity constraints are not satisfied
+                if not self.criterion.check_monotonicity(monotonic_constraint, lower_bound, upper_bound):
+                    continue
+
```

Comment:
Could you add a bool/bint variable `with_monotonic_cst` and do this check conditionally only if it is `True`?
It might already be computed in `BaseDecisionTree` when `monotinoc_cst` is filled.

Same for the other splitters.

Reply:
Otherwise, we can do:
```suggestion
                # Reject if monotonicity constraints are not satisfied
                if (
                    monotonic_constraint == 0 or not
                    self.criterion.check_monotonicity(
                        monotonic_constraint,
                        lower_bound,
                        upper_bound,
                    )
                ):
                    continue

```

Reply:
Let's try this.